### PR TITLE
Transformation Matrix for p5.Strands

### LINKED ISSUE #8953: Support transformation matrices in p5.strands
### Increasing access

If you want to use WebGL/WebGPU instancing to draw many particles, then you end up having to position the particles in a p5.strands shader rather than using the usual translate/rotate/scale functions. This gets complex pretty quickly, making it hard to take advantage of the performance opportunities of shaders. It would be great to make it as close as we can to the ease of use of regular p5 positioning.

### Most appropriate sub-area of p5.js?

- [ ] Accessibility
- [ ] Color
- [ ] Core/Environment/Rendering
- [ ] Data
- [ ] DOM
- [ ] Events
- [ ] Image
- [ ] IO
- [x] Math
- [ ] Typography
- [ ] Utilities
- [x] WebGL
- [ ] Build process
- [ ] Unit testing
- [ ] Internationalization
- [ ] Friendly errors
- [ ] Other (specify if possible)

### Feature enhancement details

Here are some examples of what it looks like right now:
- https://openprocessing.org/@davepagurek/2845379
- https://editor.p5js.org/davepagurek/sketches/Y4gEVBKaL
- https://openprocessing.org/@davepagurek/2838235

Notably, that last one uses rotations, and is really complicated!

So we can't make this look *exactly* like regular p5 transformations because we're working at a different level: normally we're moving space around ourselves and then drawing something there; in a shader, we're looking at an individual vertex at a time and updating its position/normal. 

However, I think we can do something fairly close, taking inspiration from the [web `DOMMatrix` API.](https://developer.mozilla.org/en-US/docs/Web/API/DOMMatrixReadOnly) Here's a comparison of using transforms in p5 to applying a `DOMMatrix` to a point:

<table>
<tr>
<th>

p5/shape

</th><th>

`DOMMatrix`/point

</th><th>

`DOMMatrixReadOnly`/point

</th></tr>
<tr>
<td>

```js
translate(25, 25)
rotate(-PI * 0.2)
translate(-25, -25)
scale(2)
model(something)
```

</td>
<td>

```js
const t = new DOMMatrix()
t.translateSelf(25, 25)
t.rotateSelf(36) // DOMMatrix uses degrees by default
t.translateSelf(-25, -25)
t.scaleSelf(2)
const transformed = t.transformPoint(somePoint)
```

</td>
<td>

```js
const t = new DOMMatrix()
  .translate(25, 25)
  .rotate(36) // DOMMatrix uses degrees by default
  .translate(-25, -25)
  .scale(2)
const transformed = t.transformPoint(somePoint)
```

</td>
</tr>
</table>

A matrix should ideally have these methods from `DOMMatrix`:
- `translate`
- `rotate` (and `rotateAxisAngle`)
- `scale`
- `skewX` and `skewY`
- `inverse`
- `multiply`
- `transformPoint`

Additionally, I think we also need `transformNormal`. In 3D, to make lighting work, we also need to update the orientation of the surface. This involves taking the transpose of the upper-left 3x3 submatrix of the 4x4 matrix of a 3D affine transform and then applying that to the normal vector.

### Note: to mutate or not to mutate?

I don't think we'd want to have the `*Self` suffixes like in `DOMMatrix`. I feel like we should use the un-suffixed names, but decide whether they mutate in place, or return a modified version.

Here's a little pro/con list of mutation:

<table>
<tr><th>For mutation</th><th>Against mutation</th></tr>
<tr>
<td>

- `p5.Vector` methods mutate
- Method chainining is not something we see much of in p5 code, and doing `t = t.multiply(something)` each line to avoid chaining is a little more verbose

</td>
<td>

- p5.strands vectors *don't* do any mutation in place -- they mirror GLSL vectors, where they are data like numbers in js, and global functions (not methods!) like `cross(v1, v2)` or `normalize(v)` return new vectors -- so we would be consistent within strands
- avoids a common bug where you accidentally mutate an object that you intend to reuse in another calculation

</td>
</tr>
</table>

I personally lean towards *not* mutating in p5.strands for consistency within strands. To make it less confusing with `p5.Vector`, I propose **not making methods on matrices in strands**, similar to vectors in strands, and making global methods called `translate(...)`, `scale(...)`, etc. We can reuse the same methods that in regular p5 mutate the current coordinate space, and the idea would be that because there is no global coordinate space you can mutate in strands, you use those methods slightly differently in the strands context, and then there's also no confusion in p5 about whether they'll mutate the objects, since `p5.Vector` static methods also don't mutate.

In the future, matrices outside of strands might get methods that do mutate.

### Note: matrices or transforms?

While in GLSL/WGSL these are matrices, they are used almost exclusively for transforms. I think we still want to support `mat3(...)`, `mat4(...)` etc constructors for GLSL compatibility, but as aliases for `transform2D(...)` and `transform3D(...)` respectively. (Multiplying a 3D transform, which is a `mat4`, against a `vec3`, via `transformPoint` would under the hood add a `1` to its end to make it an affine point. We can maybe let the `*` operator do "raw" multiplication.)

This terminology keeps the door open for us to have more math-oriented matrices in p5, which have different assumptions (and performance tradeoffs) than ones intended for graphics. In the future we could have a non-strands transform with a similar API (a rename of the current `p5.Matrix`, keeping its transform/graphics methods?) alongside a non-strands `p5.Matrix` too.

### PR DESCRIPTION
**Pull request by perminder-17**
Resolves #8953 

 Changes:

 Screenshots of the change:

#### PR Checklist

- [ ] `npm run lint` passes
- [ ] [Inline reference] is included / updated
- [ ] [Unit tests] are included / updated

[Inline reference]: https://p5js.org/contribute/contributing_to_the_p5js_reference/
[Unit tests]: https://github.com/processing/p5.js/tree/main/contributor_docs#unit-tests

### CODE REVIEW

File: src/strands/strands_api.js [hunk 1]
```diff
@@ -837,6 +837,186 @@ const _rgb2hsl = (instance, colorNode) => {
...
+    const [t, axis, angle] = args;
+    if (!t?.isStrandsNode || t.typeInfo().baseType !== BaseType.MAT || t.dimension !== 4) {
+      FES.userError('type error',
+        'rotateAxisAngle() only works on a 3D transform created with transform3D().');
+    }
+    const n = this.normalize(axis);
+    const x = n.x, y = n.y, z = n.z;
+    const c = this.cos(angle);
+    const s = this.sin(angle);
+    const omc = p5.strandsNode(1).sub(c); // 1 - cos
```
Comment (perminder-17) [inline src/strands/strands_api.js]: Hi @davepagurek ,

Just thinking out loud. In our non-strands modules, each area of functionality lives in its own file, for example, `transform.js` keeps the JSDoc, examples, and implementation together in one place.

I was thinking of following that same convention for the strands transforms: moving the matrix and transform helpers out of `strands_api.js` into a dedicated `strands_transform.js` (something better name for the file), with each function's docs kept alongside its implementation.

One thing to check: right now the strands functions are documented centrally in `p5.strands.js` rather than inline. Would you prefer keeping that, or colocating the transform docs with the code in the new file?

**Comment by ksen0** [comment]
Just starte looking at this one, really exciting! I will work on some more complex sketches to test it out, but I appreciate the docs you've included out of the box. 

Would it be a good idea to add some bench and/or unit and/or visual tests here? I think visual and unit at minimum, given the scope of the PR, would be good...

**Comment by perminder-17** [comment]
> Just starte looking at this one, really exciting! I will work on some more complex sketches to test it out, but I appreciate the docs you've included out of the box.

Ah, correct. I'm also getting feedback and testing it with more complex examples myself. For now, I'm trying to find ways to break the feature and catch any issues. 

Once I'm confident with the code, I'll add tests for them.

File: src/strands/strands_transform.js [hunk 2]
```diff
@@ -0,0 +1,919 @@
...
+   * let myShader;
+   *
+   * function setup() {
+   *   createCanvas(200, 200, WEBGL);
+   *   myShader = buildMaterialShader(slide);
+   * }
+   *
+   * function slide() {
+   *   getWorldInputs((inputs) => {
+   *     // translate() returns a new transform, so assign the result back.
```
Comment (davepagurek) [inline src/strands/strands_transform.js:348]: Maybe add an inline URL here? Or add a note at the end of the `transform()` docs to mention strands + link to the `transformPoint` docs as the spot with more in depth explanation

Comment (davepagurek) [reply src/strands/strands_transform.js:348]: Or interleave some writing before the strands example saying the same thing

File: src/strands/strands_transform.js [hunk 3]
```diff
@@ -0,0 +1,919 @@
...
+   * @example
+   * let myShader;
+   *
+   * function setup() {
+   *   createCanvas(200, 200, WEBGL);
+   *   myShader = buildMaterialShader(wobble);
+   * }
+   *
+   * function wobble() {
+   *   getWorldInputs((inputs) => {
```
Comment (davepagurek) [inline src/strands/strands_transform.js:147]: Can we use `worldInputs.begin()` and `worldInputs.end()` in the examples for consistency?

**Comment by davepagurek** [comment]
Some feedback:
- We could make visual tests out of examples
- We should make sure that in the website reference, we don't accidentally recategorize the transform methods into the strands category. Maybe we need to add e.g. `@module` tags onto individual methods?
- We also want the strands parts to come last, it may be that adding our docs to the existing docs rather than making new ones and relying on the docs builder to stitch them together is the most reliable way of doing that
- Once main is merged in, could we add examples using `instances(n)` and using transforms per instance?
- Not blocking for this PR but thoughts for the future: for sketches like https://openprocessing.org/@davepagurek/2985673 when transforms vary based on position, the current `transformNormal` logic is not quite correct. That sketch has some code to figure out what the new normal should be by sampling a few nearby points, but it relies on a slightly different structure (your transform is a function based on an input position, not just a single matrix.) Any thoughts on what it could look like in the future if we wanted to support that? Just to confirm that it wouldn't clash with our current API?
- Just noticed that the current p5 transforms for skewing are called `shearX` and `shearY` but it looks like the ones we've added here use `skew` as a prefix instead, based on the DOMMatrix API. I hadn't realized we use different conventions there. Do you or @ksen0 have thoughts on whether it's better to just use what p5 already has for naming there?

**Comment by davepagurek** [comment]
oh and one more thing that I forgot: add on to the logging system for experimental features in https://github.com/processing/p5.js/pull/9109. It sounds like we don't want a whole new contributor doc, but we just want to add a SECTION to the strands doc, and then link to that section via a `#sectionname` suffix (I think we do that in tables of contents in other contributor docs.) We may need to update the system a little to make a category that links to a section of a doc rather than to a whole doc of its own, + we may need to use the manual checker rather than the decorator to catch uses of these functions since we don't want it logging in regular non-strands use
